### ISSUE TITLE
[css-grid-2] Resolving padding in indefinite intrinsic size calculation and using that to resolve indefinite size?

### ISSUE THREAD

**Issue by KurtCattiSchmidt** [w3c/csswg-drafts#8963]
[Spec](https://www.w3.org/TR/css-grid-2/#intrinsic-sizes)

I'm looking at [grid-auto-repeat-max-size-001.html](http://wpt.live/css/css-grid/grid-definition/grid-auto-repeat-max-size-001.html) in order to fix [this Compat 2023 bug in Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=1449940)

...however, it appears that both Blink and WebKit make more sense to me than what Gecko is doing, although Gecko's behavior is marked as passing.

Note that this is in relation to sub-cases 16 and 19, which Blink and WebKit are marked as failing and Gecko is marked as passing. The other cases are all in agreement.

I'll explain what Blink and WebKit are doing, and why I think that is the proper behavior.

No width or height are specified on the grid, so for sizing purposes, we fall back to the default intrinsic size of 300px by 150px. However, with this fallback, we are able to determine a final size of the grid by the end of the grid layout algorithm that differs from 300px by 150px.

Padding is 10%, on an indefinite inline axis, which falls back to 300px, so we land at 30px padding. This is agreed upon in all implementations. 

Grid-template is set to `repeat(auto-fill, 50px) / repeat(auto-fill, 100px);` which we can resolve to two repetitions in both directions to fill the indefinite 300px width and 150px indefinite height (it can't be 3 repetitions in either axis due to the padding). This is also agreed upon in all implementations.

This is where things diverge in Gecko. For the inline size, Blink/Webkit add up the 30px padding on each side, along with 2 repetitions of 100px for each column, and land at 260px. Gecko sticks with the default 300px intrinsic size, even though it is possible to calculate an intrinsic size based on the fixed column sizes. There seems to be some extra padding *after* the grid items in Gecko, which is how it lands at 300px. I can't find any reference to this in the grid spec.

For block size, interestingly, Blink/WebKit/Gecko all agree on adding up padding-top and padding-bottom (30px + 30px) along with two rows of 50px, landing at 160px. 

It's not clear to my why Gecko is inconsistent between the inline and block direction here. The grid spec implies that if fixed sizes are specified on track sizes (as is happening here), that those will be used for intrinsic size calculation.

If there is something I'm missing here, and that Gecko is indeed doing this correctly, I'd be happy to update Chromium's implementation. Otherwise, I'll update the expected sizes in the test case.

@bfgeek created a reduced case that demonstrates the extra padding Gecko is adding: https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11785

**Comment by bfgeek** [comment]
@dholbert - Do you know where this additional padding is coming from? (We are super confused).
Testcase:

```
<!DOCTYPE html>
<div style="width: 300px; border: solid 3px;">
  <div style="
    background: gray;
    box-sizing: border-box;
    display: grid;
    padding: 10%;
    width: min-content;
    max-width: 300px;
    grid-template-columns: repeat(auto-fill, 100px);
    grid-template-rows: 100px;">
      <div style="
        background: lime;
        grid-column: -2;"></div>
  </div>
</div>

<div style="width: 300px; border: solid 3px;">
  <div style="
    background: gray;
    box-sizing: border-box;
    display: grid;
    padding: 10%;
    width: min-content;
    max-width: 300px;
    grid-template-columns: repeat(2, 100px);
    grid-template-rows: 100px;">
      <div style="
        background: lime;
        grid-column: -2;"></div>
  </div>
</div>
```

(only difference is `auto-fill` vs. `2` in `grid-template-columns`) which resolve to the same thing in both cases.

**Comment by Loirooriol** [comment]
> we fall back to the default intrinsic size of 300px by 150px

This doesn't make sense to me, the grid is not a replaced element. https://drafts.csswg.org/css-sizing-3/#intrinsic-sizes says that 300x150 is for "replaced elements without natural sizes".

AFAIK what should happen is basically this:
 - 1st we need to know the intrinsic contribution of the grid container. So we resolve `repeat(auto-fill, 100px)`. Per https://www.w3.org/TR/css-grid-2/#auto-repeat it should be 3 repetitions, using the `max-width`. So we run the track sizing algorithm under a min/max-content constraint, we get 3 tracks of 100px each and no gap, so the grid is 300px.
 - Then we use the 300px as the size of the content area of the grid container. There is also a `padding: 10%`, but per https://drafts.csswg.org/css-sizing-3/#min-percentage-contribution it resolves to 0.
 - So the intrinsic contribution is 300px, and the parent element is also sized to 300px.
 - Then we lay out the grid container "for real".
 - We can now resolve `padding: 10%` against the parent, so padding becomes 30px.
 - This means that the border area would be 360px, but we have `max-width: 300px; box-sizing: border-box`, so the content area shrinks to 240px so that the border area doesn't exceed 300px.
 - Now when resolving `repeat(auto-fill, 100px)` we can only have 2 repetitions without overflowing the content area. So we get 2 tracks of 100px each.

I think it should be analogous in the block axis, maybe Firefox does a different thing since it doesn't implement the 2nd pass (https://bugzilla.mozilla.org/show_bug.cgi?id=1481876).

**Comment by Loirooriol** [comment]
In other words, I think it's Blink the one that adds some weird spacing between the grid and its parent in this [case](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cdiv%20style%3D%22width%3A%20min-content%3B%20border%3A%20solid%203px%3B%22%3E%0A%20%20%3Cdiv%20style%3D%22%0A%20%20%20%20background%3A%20gray%3B%0A%20%20%20%20box-sizing%3A%20border-box%3B%0A%20%20%20%20display%3A%20grid%3B%0A%20%20%20%20padding%3A%2010%25%3B%0A%20%20%20%20width%3A%20min-content%3B%0A%20%20%20%20max-width%3A%20200px%3B%0A%20%20%20%20grid-template-columns%3A%20repeat(auto-fill%2C%20100px)%3B%0A%20%20%20%20grid-template-rows%3A%20100px%22%3E%0A%20%20%20%20%20%20%3Cdiv%20style%3D%22background%3A%20lime%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E):

```html
<!DOCTYPE html>
<div style="width: min-content; border: solid 3px;">
  <div style="
    background: gray;
    box-sizing: border-box;
    display: grid;
    padding: 10%;
    width: min-content;
    max-width: 200px;
    grid-template-columns: repeat(auto-fill, 100px);
    grid-template-rows: 100px">
      <div style="background: lime"></div>
  </div>
</div>
```

**Comment by bfgeek** [comment]
> Then we use the 300px as the size of the content area of the grid container. There is also a padding: 10%, but per https://drafts.csswg.org/css-sizing-3/#min-percentage-contribution it resolves to 0.


This is likely the crux of the issue. I think this part of the spec is incorrect and likely should change this. Blink/Webkit will resolve percentages if they are resolvable (e.g. we'll resolve the %-padding for the top-most container). And if un-resolvable - they'll resolve against zero.

E.g. this yeilds better results for cases like:
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11788

(In firefox you'll end up with something not-square for the above case).
%-padding on the container almost never matters, except for height resolution, and now in grid as its resolving min/max widths.

**Comment by KurtCattiSchmidt** [comment]
> > we fall back to the default intrinsic size of 300px by 150px
> 
> This doesn't make sense to me, the grid is not a replaced element. https://drafts.csswg.org/css-sizing-3/#intrinsic-sizes says that 300x150 is for "replaced elements without natural sizes".
> 
> AFAIK what should happen is basically this:
> 
> * 1st we need to know the intrinsic contribution of the grid container. So we resolve `repeat(auto-fill, 100px)`. Per https://www.w3.org/TR/css-grid-2/#auto-repeat it should be 3 repetitions, using the `max-width`. So we run the track sizing algorithm under a min/max-content constraint, we get 3 tracks of 100px each and no gap, so the grid is 300px.
> * Then we use the 300px as the size of the content area of the grid container. There is also a `padding: 10%`, but per https://drafts.csswg.org/css-sizing-3/#min-percentage-contribution it resolves to 0.
> * So the intrinsic contribution is 300px, and the parent element is also sized to 300px.
> * Then we lay out the grid container "for real".
> * We can now resolve `padding: 10%` against the parent, so padding becomes 30px.
> * This means that the border area would be 360px, but we have `max-width: 300px; box-sizing: border-box`, so the content area shrinks to 240px so that the border area doesn't exceed 300px.
> * Now when resolving `repeat(auto-fill, 100px)` we can only have 2 repetitions without overflowing the content area. So we get 2 tracks of 100px each.
> 
> I think it should be analogous in the block axis, maybe Firefox does a different thing since it doesn't implement the 2nd pass (https://bugzilla.mozilla.org/show_bug.cgi?id=1481876).

The discrepancy here is actually in the inline size, where Blink/WebKit resolve to 240px as you describe, and Firefox is the one at 300px. After reading your breakdown, I believe I was incorrect in how some of these values are computed and that you described it more accurately. 

In the block axis, Firefox matches what Blink/Webkit are doing in the inline axis. I don't think it's the second-pass logic as you described. 

Are we in agreement that this particular test case expectations are wrong, and that the inline sizes for cases 16 and 19 should be 240px instead of 300px? If so, I can get a PR ready to update it.

**Comment by Loirooriol** [comment]
> Blink/WebKit resolve to 240px as you describe

Now I'm very confused, in the test I'm getting this:

| | Firefox  | Chromium |
| ------------- | ------------- | ------------- |
| Wrapper  | Content area: 300x160<br>Padding area: 300x160  | Content area: 300x160<br>Padding area: 300x160  | 
| Grid container  | Content area: 240x100<br>Padding area: 300x160 | Content area: 200x100<br>Padding area: 260x160  |

> Are we in agreement that this particular test case expectations are wrong

No, I agree with `data-expected-width="300"` (it checks `offsetWidth`, including the padding).

**Comment by Loirooriol** [comment]
> E.g. this yeilds better results for cases like: https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11788

Here the `padding: 10%` is not cyclic, the container has a definite size, so I think it's something different.

> %-padding on the container almost never matters, except for height resolution, and now in grid as its resolving min/max widths

I don't think this is something new, I think these should behave the same (like they do on Firefox): https://www.software.hixie.ch/utilities/js/live-dom-viewer/saved/11789

```html
<!DOCTYPE html>
<style>
.wrapper { width: min-content; border: solid 3px }
.target { background: gray; box-sizing: border-box; padding: 10%; width: min-content; max-width: 200px; }
.target.grid { display: grid; grid-template-columns: repeat(auto-fill, 100px); grid-template-rows: 100px }
.target::before { content: ""; display: block; background: lime }
.target.block::before { width: 200px; max-width: max(-1%, 100px); height: 100px }
</style>
<div class="wrapper">
  <div class="target block"></div>
</div>
<div class="wrapper">
  <div class="target grid"></div>
</div>
```

In both cases we have some contents with a width of 200px during intrinsic sizing which shrinks to 100px when laying out for real. But the padding area of `.target.block` keeps being 200px, I son't see why it should be different for `.target-grid`.

**Comment by KurtCattiSchmidt** [comment]
> > Blink/WebKit resolve to 240px as you describe
> 
> Now I'm very confused, in the test I'm getting this:
> 
> Firefox	Chromium
> Wrapper	Content area: 300x160
> Padding area: 300x160	Content area: 300x160
> Padding area: 300x160
> Grid container	Content area: 240x100
> Padding area: 300x160	Content area: 200x100
> Padding area: 260x160
> > Are we in agreement that this particular test case expectations are wrong
> 
> No, I agree with `data-expected-width="300"` (it checks `offsetWidth`, including the padding).

Sorry, I meant 260px. This is the testharness.js output:

.grid 16

assert_equals: <div class="grid border border-box pad-percent" style="width: min-content; height: min-content;" data-expected-width="300" data-expected-height="160"> <div class="item" data-offset-x="130" data-offset-y="80" data-expected-width="100" data-expected-height="50"></div> </div> width expected 300 but got 260

.grid 19

assert_equals: <div class="grid border border-box pad-percent" style="width: max-content; height: max-content;" data-expected-width="300" data-expected-height="160"> <div class="item" data-offset-x="130" data-offset-y="80" data-expected-width="100" data-expected-height="50"></div> </div> width expected 300 but got 260

**Comment by Loirooriol** [comment]
> I don't think it's the second-pass logic as you described.

Yeah, I guess the difference is because when finding the intrinsic size in the block axis, Firefox already knows the width of the wrapper so it can resolve the percentage. That is, the asymmetry would be because only the horizontal paddings are cyclic.

**Comment by bfgeek** [comment]
> I don't think this is something new, I think these should behave the same (like they do on Firefox): https://www.software.hixie.ch/utilities/js/live-dom-viewer/saved/11789

These two tests are pretty different, and aren't really comparable.

> That is, the asymmetry would be because only the horizontal paddings are cyclic.

I don't think the %-padding is cyclic once inside the layout pass of the inner grid, you have a definite width once you work out what the outer "min-content" resolves to, and can resolve the %-padding - which in turn will alter the min-content size of the inner grid. This happens quite often.

This is similar to the 2nd pass logic for height in grid. This will change the min-content size of a box.

**Comment by bfgeek** [comment]
E.g. concretely: https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11791

Even though the boxes are sized by `width: max-content` we still resolve the %-padding during the layout pass. Grid shouldn't be different here?

**Comment by Loirooriol** [comment]
Yeah, resolving `padding: 10%` can increase the outer size and make the element overflow, but it shouldn't affect `width: min-content`.

Grid in Blink can contribute 300px to the parent, but resolve `width: min-content` to an outer size of 260px, seems strange.

**Comment by bfgeek** [comment]
> Grid in Blink can contribute 300px to the parent, but resolve width: min-content to an outer size of 260px, seems strange.

But this is just the padding resolving and causing the repeat() to evaluate differently. :). It might be strange, but unclear how its incorrect?

**Comment by bfgeek** [comment]
Additionally if we modify your testcase to make the outer div definite, e.g. the padding is definitely not cyclic, Firefox doesn't respect it:
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11792

**Comment by Loirooriol** [comment]
Yes the `repeat()` should change, but not the grid container. Just like here:

```html
<div style="width: min-content; border: solid 3px">
  <div style="background: cyan; display: grid; width: min-content; grid-template-columns: 0%">
    loremipsum
  </div>
</div>
```

During intrinsic sizing, `0%` behaves as `auto` and the column is as wide as the text. This fixes the width of the grid container and the wrapper. Then we can resolve `0%` to `0px`, but the grid container is still as wide as the text. It's not resized to the new size of the grid.

So why does Blink resize the grid container in other cases?

**Comment by Loirooriol** [comment]
> Additionally if we modify your testcase to make the outer div definite, e.g. the padding is definitely not cyclic

Yeah Firefox seems wrong here

**Comment by bfgeek** [comment]
> So why does Blink resize the grid container in other cases?

Because the padding now *can* be resolved, as we have a definite width to resolve against.

> Yeah Firefox seems wrong here https://searchfox.org/mozilla-central/rev/f76d80e486028313488d05f4e7fe2509cec11777/layout/generic/nsGridContainerFrame.cpp#204-206

This is kinda my point. E.g. https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11797

```
<div class="wrapper" style="width: min-content;">
  <div class="target grid"></div>
</div>
<div class="wrapper" style="width: 200px;">
  <div class="target grid"></div>
</div>
```

Once `width: min-content` is resolved and definite - then it should be no different to if `width: 200px` was specified. Boxes above should render exactly the same.

**Comment by bfgeek** [comment]
A more straightforward example might be:

```
<!DOCTYPE html>
<div style="width: min-content; background: lime;">
  <div style="width: 0%; outline: solid red;">
    abc
  </div>
</div>
```

Here the width of the inner div doesn't stay as "auto" it'll re-resolve as `0px`. The same is true for padding.

**Comment by Loirooriol** [comment]
It still doesn't make sense to me that Blink behaves different in [these cases](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A.wrapper%20%7B%20width%3A%20max-content%3B%20border%3A%20solid%203px%3B%20%7D%0A.grid%20%7B%20background%3A%20gray%3B%20display%3A%20grid%3B%20grid-template-rows%3A%20100px%3B%0A%20%20%20%20%20%20%20%20box-sizing%3A%20border-box%3B%20padding%3A%2010%25%3B%20width%3A%20min-content%3B%20max-width%3A%20200px%3B%20%7D%0A.grid%3A%3Abefore%20%7B%20content%3A%20%22%22%3B%20background%3A%20lime%3B%20width%3A%20100px%20%7D%0A.grid%3A%3Aafter%20%7B%20content%3A%20%22%22%3B%20background%3A%20lime%3B%20width%3A%20100px%20%7D%0A%3C%2Fstyle%3E%0A%3Cdiv%20class%3D%22wrapper%22%3E%0A%20%20%3Cdiv%20class%3D%22grid%22%20style%3D%22grid-template-columns%3A%20repeat(auto-fill%2C%2051%25)%22%3E%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20class%3D%22wrapper%22%3E%0A%20%20%3Cdiv%20class%3D%22grid%22%20style%3D%22grid-template-columns%3A%20repeat(auto-fill%2C%20100px)%22%3E%3C%2Fdiv%3E%0A%3C%2Fdiv%3E):

```html
<!DOCTYPE html>
<style>
.wrapper { width: max-content; border: solid 3px; }
.grid { background: gray; display: grid; grid-template-rows: 100px;
        box-sizing: border-box; padding: 10%; width: min-content; max-width: 200px; }
.grid::before { content: ""; background: lime; width: 100px }
.grid::after { content: ""; background: lime; width: 100px }
</style>
<div class="wrapper">
  <div class="grid" style="grid-template-columns: repeat(auto-fill, 51%)"></div>
</div>
<div class="wrapper">
  <div class="grid" style="grid-template-columns: repeat(auto-fill, 100px)"></div>
</div>
```

**Comment by dholbert** [comment]
Replying to the following (a bit too late since @Loirooriol beat me to an answer):
> @dholbert - Do you know where this additional padding is coming from? (We are super confused).

As Oriol noted, it boils down to the line that he quoted in https://github.com/w3c/csswg-drafts/issues/8963#issuecomment-1590243087 , which is where we determine the space that we have available for `repeat(auto-fill, 100px)` to expand into.

So we end up thinking that we can fit 3 columns, when we're computing the intrinsic size of the grid.

(Need to think some more about comparable scenarios & to-what-extent this is broken...)

**Comment by bfgeek** [comment]
> It still doesn't make sense to me that Blink behaves different in [these cases](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A.wrapper%20%7B%20width%3A%20max-content%3B%20border%3A%20solid%203px%3B%20%7D%0A.grid%20%7B%20background%3A%20gray%3B%20display%3A%20grid%3B%20grid-template-rows%3A%20100px%3B%0A%20%20%20%20%20%20%20%20box-sizing%3A%20border-box%3B%20padding%3A%2010%25%3B%20width%3A%20min-content%3B%20max-width%3A%20200px%3B%20%7D%0A.grid%3A%3Abefore%20%7B%20content%3A%20%22%22%3B%20background%3A%20lime%3B%20width%3A%20100px%20%7D%0A.grid%3A%3Aafter%20%7B%20content%3A%20%22%22%3B%20background%3A%20lime%3B%20width%3A%20100px%20%7D%0A%3C%2Fstyle%3E%0A%3Cdiv%20class%3D%22wrapper%22%3E%0A%20%20%3Cdiv%20class%3D%22grid%22%20style%3D%22grid-template-columns%3A%20repeat(auto-fill%2C%2051%25)%22%3E%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20class%3D%22wrapper%22%3E%0A%20%20%3Cdiv%20class%3D%22grid%22%20style%3D%22grid-template-columns%3A%20repeat(auto-fill%2C%20100px)%22%3E%3C%2Fdiv%3E%0A%3C%2Fdiv%3E):

The difference here is that the "51%" gets resolved against zero during the min/max pass, and so will compute the number of repetitions to be 200, (search for: "It is suggested that this floor be 1px" in the spec). The spec is kinda vauge in this area and doesn't explicitly call out what to resolve the percentages against.

As we have 200 columns we'll place the two children beside each other get get a min/max content size 200px.
Once inside the layout pass we can resolve the percentages, and only fit a single column.

> (Need to think some more about comparable scenarios & to-what-extent this is broken...)

@dholbert Do you agree Firefox seems wrong for this case? https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11792

**Comment by Loirooriol** [comment]
> Once inside the layout pass we can resolve the percentages, and only fit a single column.

This happens on both cases. Please explain why the grid container should only be resized in one of them.

**Comment by Loirooriol** [comment]
> The padding percentage is resolved against the grid's containing-block (which is definite when determining the min/max-content size of the grid)

How is it definite if it has `.wrapper { width: max-content;`? The padding is also cyclic so is treated as 0 when determining the min/max-content size of the grid.

**Comment by bfgeek** [comment]
> How is it definite if it has .wrapper { width: max-content;? The padding is also cyclic so is treated as 0 when determining the min/max-content size of the grid.

The padding resolves against the containing-block. When determining the width of the `wrapper` it is resolved against zero - as the containing-block doesn't have a size yet. When determining the width of `grid` (we don't determine the final width of `grid` until *after* we've determined a final size of `wrapper`), we can resolve the padding against the containing-block. It doesn't matter that `wrapper` has `width: max-content` applied. After we've determined the final width of `wrapper` it should behave the same as if it had `width: Npx` applied to it.

This is the point being made in comment: https://github.com/w3c/csswg-drafts/issues/8963#issuecomment-1591968925

**Comment by Loirooriol** [comment]
> we don't determine the final width of grid until after we've determined a final size of wrapper

Sure, for the final width, but for the min/max-content size the order should be the opposite: first the grid, then the wrapper. So when determining the min/max-content size of the grid, we still don't know the size of the wrapper, and then can't resolve the padding percentage.

**Comment by bfgeek** [comment]
> Sure, for the final width, but for the min/max-content size the order should be the opposite: first the grid, then the wrapper. So when determining the min/max-content size of the grid, we still don't know the size of the wrapper, and then can't resolve the padding percentage.

There isn't "one" min/max-content size for a box. It depends on the state from the containing-block. E.g. what to resolve the padding against, and what to resolve the block-size against.

**Comment by dholbert** [comment]
> @dholbert Do you agree Firefox seems wrong for this case? https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11792

Sorry, forgot about this thread. Yes, I agree that Firefox seems wrong for that case, and I think your logic in https://github.com/w3c/csswg-drafts/issues/8963#issuecomment-1595379623 makes sense.
